content/kb/tutorials/deploy/docker.md

```diff
@@ -0,0 +1,236 @@
+
+Here's an example `Dockerfile` that you can add to the root of your directory. i.e. in `app/`
+
+```dockerfile
+# app/Dockerfile
+
+FROM python:3.9-slim
+
+EXPOSE 8501
+
+WORKDIR /app
+
+RUN apt-get update && apt-get install -y \
+    build-essential \
+    software-properties-common \
+    git \
+    && rm -rf /var/lib/apt/lists/*
+
+RUN git clone https://github.com/streamlit/streamlit-example.git .
+
+RUN pip3 install -r requirements.txt
+
+ENTRYPOINT ["streamlit", "run", "streamlit_app.py", "--server.port=8501", "--server.address=0.0.0.0"]
+```
+
+### Dockerfile walkthrough
+
+Let’s walk through each line of the Dockerfile :
+
+1. A `Dockerfile` must start with a [`FROM`](https://docs.docker.com/engine/reference/builder/#from) instruction. It sets the [Base Image](https://docs.docker.com/glossary/#base-image) (think OS) for the container:
+
+   ```dockerfile
+   FROM python:3.9-slim
+   ```
+
+   Docker has a number of official Docker base images based on various Linux distributions. They also have base images that come with language-specific modules such as [Python](https://hub.docker.com/_/python). The `python` images come in many flavors, each designed for a specific use case. Here, we we use the `python:3.9-slim` image which is a lightweight image that comes with the latest version of Python 3.9.
+
+   <Tip>
+
+   You can also use your own base image, provided the image you use contains a [supported version of Python](/knowledge-base/using-streamlit/sanity-checks#check-0-are-you-using-a-streamlit-supported-version-of-python) for Streamlit. There is no one-size-fits-all approach to using any specific base image, nor is there an official Streamlit-specific base image.
+
+   </Tip>
+
+2. The [`EXPOSE`](https://docs.docker.com/engine/reference/builder/#expose) instruction informs Docker that the container listens on the specified network ports at runtime. Your container needs to listen to Streamlit’s (default) port 8501:
+
+   ```dockerfile
+   EXPOSE 8501
+   ```
+
+3. The `WORKDIR` instruction sets the working directory for any `RUN`, `CMD`, `ENTRYPOINT`, `COPY` and `ADD` instructions that follow it in the `Dockerfile` . Let’s set it to `app/` :
+
+   ```dockerfile
+   WORKDIR /app
+   ```
+
+4. Install `git` so that we can clone the app code from a remote repo:
+
+   ```dockerfile
+   RUN apt-get update && apt-get install -y \
+       build-essential \
+       software-properties-common \
+       git \
+       && rm -rf /var/lib/apt/lists/*
+   ```
+
+5. Clone your code that lives in a remote repo to `WORKDIR`:
+
+   ```dockerfile
+   RUN git clone https://github.com/streamlit/streamlit-example.git .
```

Comment:
I'm afraid most users will be using private repositories, so we should describe how to install applications from a local drive or [using ssh to access private repository](https://docs.docker.com/develop/develop-images/build_enhancements/#using-ssh-to-access-private-data-in-builds) without copy private key to repository

Reply:
The instructions in Notion intially described how to `COPY` Streamlit application code from the local filesystem to the container. That was removed in favor of git cloning. We could however break it down by 3 steps:
- Clone a public repo
- Close a private repo with ssh without copying the priv key the repo
- Copying the local app dir into the container

But I'm not 100% convinced we should basically re-write GitHub's [using ssh to access private repository](https://docs.docker.com/develop/develop-images/build_enhancements/#using-ssh-to-access-private-data-in-builds) instructions. Should we just link to it or re-write ourselves? WDYT @randyzwitch?

Reply:
On second thought, perhaps we should include private repo instructions for completeness, especially if believe the avg user of the tutorial is cloning a private repo (?)

Reply:
I'm not sure completeness needs to be a desired goal, all of these tutorials IMO should be sketches of how things can be done. If we give people an example that "if you get your code inside of the Docker container, then do...", if people have their own quirks then they are hopefully qualified to alter the example to fit their needs.